docs/next_link_analysis.md

```diff
@@ -0,0 +1,108 @@
+
+A particular advantage of this structure is that it easily handles the high cardinality of the data (many different potential values for the target). It can also be built online.
+
+### Scoring test data
+
+In order to get the rank of an observation in the test data, the tuple from the test data looks up the rank in the train data.
+
+In practice, if this solution was deployed, the first part of the tuple key would be used (ie the current page), and the top [10] predictions would be returned as a list of `key:value` pairs.
+
+### Illustration of train dictionary using Companies House
+
+The most popular URL is for Companies House. This is a view of the dictionary of links for Companies House. Due to the dictionary structure, the entries are not sorted by value. The values represent the count of frequency of the second part of the key in the train dataset. In this example, the most popular link is to `get-information-about-a-company`.
+
+![train dictionary](./images/train_dictionary.png)
+
+### Illustration of results by parent URL
+
+In the test set, we can aggregate by the parent URL, and find the mean reciprocal rank overall. As can be seen the results vary widely by the parent URL. The mean reciprocal rank for the companies house URL is very high, nearly 1, which reflects the dominance shown above in the category `get-information-about-a-company`.
+
+In the same way someone should not be impressed with 50% accuracy in coin toss predictions, we should not be impressed with these results. We have just quantified the natural ease of predicting the next link on GOVUK.
+
+![aggregated results](./images/results_by_parent_URL.png)
+
+In contrast to the results for the `companies-house` URL, we see that the predictive performance of 'find-local-council` is poor. This reflects the high cardinality of the potential child URL. This would be partially remedied by the inclusion of location in the explanatory data.
+
+## Final words
+
+The predictive power of the model should be seen as establishing an initial benchmark for the mean reciprocal rank (MRR). Other data should be included, most likely location and then ideally user specific information such as other URL activity over preceding sessions. In turn, the URL activity from previous sessions could be expressed by translating the related content into document vectors, for example using the average of the sentence embeddings, and including dwell times to weight the content. The benefit of additional data needs to be weighed up against the cost of acquiring the data and managing the computational environment necessary to process the data.
```

Comment:
Dwell times is an interesting one. Tim has made good suggestions before about how misleading this can be because a user may appear to spend long on a webpage so one would naturally infer that the page is relevant, when in reality, that user might have skipped to another tab to search for something else, maybe because the link they clicked was not completely relevant. It's better than nothing though.

Reply:
Yes, let the algorithm determine whether dwell time is unimportant. We may well find that its still predictive, but has some non linear/bi modal behaviour due to sets of behaviours a) someone is reading the page and flicking through pages trying to actively find information b) someone opened their browser on a page and then lost focus